src/modules/silentpayments/main_impl.h

```diff
@@ -7,10 +7,296 @@
+    int overflow = 0;
+    int ret;
+
+    /* Sanity check inputs. */
+    VERIFY_CHECK(ctx != NULL);
+    ARG_CHECK(secp256k1_ecmult_gen_context_is_built(&ctx->ecmult_gen_ctx));
+    ARG_CHECK(generated_outputs != NULL);
+    ARG_CHECK(recipients != NULL);
+    ARG_CHECK(n_recipients > 0);
+    ARG_CHECK((plain_seckeys != NULL) || (taproot_seckeys != NULL));
+    if (taproot_seckeys != NULL) {
+        ARG_CHECK(n_taproot_seckeys > 0);
+    } else {
+        ARG_CHECK(n_taproot_seckeys == 0);
+    }
+    if (plain_seckeys != NULL) {
+        ARG_CHECK(n_plain_seckeys > 0);
+    } else {
+        ARG_CHECK(n_plain_seckeys == 0);
+    }
+    ARG_CHECK(outpoint_smallest36 != NULL);
+    /* ensure the index field is set correctly */
+    for (i = 0; i < n_recipients; i++) {
+        ARG_CHECK(recipients[i]->index == i);
+    }
+
+    /* Compute input private keys sum: a_sum = a_1 + a_2 + ... + a_n */
+    a_sum_scalar = secp256k1_scalar_zero;
+    for (i = 0; i < n_plain_seckeys; i++) {
+        ret = secp256k1_scalar_set_b32_seckey(&addend, plain_seckeys[i]);
+        secp256k1_declassify(ctx, &ret, sizeof(ret));
+        if (!ret) {
+            secp256k1_scalar_clear(&addend);
+            secp256k1_scalar_clear(&a_sum_scalar);
+            return 0;
+        }
+        secp256k1_scalar_add(&a_sum_scalar, &a_sum_scalar, &addend);
+    }
+    /* private keys used for taproot outputs have to be negated if they resulted in an odd point */
+    for (i = 0; i < n_taproot_seckeys; i++) {
+        secp256k1_ge addend_point;
+        ret = secp256k1_keypair_load(ctx, &addend, &addend_point, taproot_seckeys[i]);
+        secp256k1_declassify(ctx, &ret, sizeof(ret));
+        if (!ret) {
+            secp256k1_scalar_clear(&addend);
+            secp256k1_scalar_clear(&a_sum_scalar);
+            return 0;
+        }
+        secp256k1_declassify(ctx, &ret, sizeof(ret));
+        if (secp256k1_fe_is_odd(&addend_point.y)) {
+            secp256k1_scalar_negate(&addend, &addend);
+        }
+        secp256k1_scalar_add(&a_sum_scalar, &a_sum_scalar, &addend);
+    }
+    /* If there are any failures in loading/summing up the secret keys, fail early */
+    ret = secp256k1_scalar_is_zero(&a_sum_scalar);
+    secp256k1_declassify(ctx, &ret, sizeof(ret));
+    if (ret) {
+        secp256k1_scalar_clear(&addend);
+        secp256k1_scalar_clear(&a_sum_scalar);
+        return 0;
+    } else {
+        /* Clear the addend variable as this is no longer needed at this point
+         * and contains secret data. This saves from needing to remember to clear
+         * this variable from multiple places below */
+        secp256k1_scalar_clear(&addend);
+    }
+    /* Compute input_hash = hash(outpoint_L || (a_sum * G)) */
+    secp256k1_ecmult_gen(&ctx->ecmult_gen_ctx, &A_sum_gej, &a_sum_scalar);
+    secp256k1_ge_set_gej(&A_sum_ge, &A_sum_gej);
+    /* TODO: comment */
+    secp256k1_declassify(ctx, &A_sum_ge, sizeof(A_sum_ge));
```

Comment:
Suggestion for the comment to tackle another TODO 🔫  (IIUC, the non-constant timing of group element serialization seems to ultimately be caused by [variable-time normalization of field elements](https://github.com/bitcoin-core/secp256k1/blob/b9313c6e1a6082a66b4c75777e18ca4b176fcf9d/src/eckey_impl.h#L44-L45)):
```suggestion
    /* Need to declassify the pubkey sum because serializing a group element (done in the
       `_calculate_input_hash` call following) is not a constant-time operation */
    secp256k1_declassify(ctx, &A_sum_ge, sizeof(A_sum_ge));
```

Reply:
Thanks for digging into this! Added the comment with some slight formatting changes.